crates/cubecl-wgpu/src/compiler/base.rs

```diff
@@ -0,0 +1,27 @@
+use std::sync::Arc;
+
+use cubecl_core::{
+    prelude::CompiledKernel, server::ComputeServer, Compiler, ExecutionMode, Feature,
+};
+use cubecl_runtime::DeviceProperties;
+use wgpu::{Adapter, ComputePipeline, Device, Queue};
+
+use crate::WgpuServer;
+
+pub trait WgpuCompiler: Compiler {
+    fn compile(
+        server: &mut WgpuServer<Self>,
+        kernel: <WgpuServer<Self> as ComputeServer>::Kernel,
+        mode: ExecutionMode,
+    ) -> CompiledKernel<Self>;
+
+    fn create_pipeline(
+        server: &mut WgpuServer<Self>,
+        kernel: CompiledKernel<Self>,
+        mode: ExecutionMode,
+    ) -> Arc<ComputePipeline>;
+
+    #[expect(async_fn_in_trait, reason = "Only used locally")]
+    async fn request_device(adapter: &Adapter) -> (Device, Queue);
+    fn register_features(adapter: &Adapter, device: &Device, props: &mut DeviceProperties<Feature>);
```

Comment:
I'm not familiar with `#[expect]`, what does it do? Also, would it be better to have only one function `request_device` that comes with all features registered?

```rust
async fn request_device(adapter: &Adapter) -> (Device, Queue, DeviceProperties<Features>);
```

Reply:
`expect` was added a few rust versions ago, I'll have to actually check if it fits with the MSRV. Might just do `allow` for now. It's `allow` but warns you if the lint is not actually needed, for example because things change and it becomes unnecessary to have it there.

Reply:
Looks like it's 1.81 MSRV, but burn is currently set to 1.80 for its minimum version. Same goes for the lint reason which was added at the same time. Should I make it an `allow` with no reason for now?

Reply:
As for merging the functions, it's not possible without changing the inialization order `wgpu` uses right now. `request_device` is used in `select_device`, but the features are set on the `Client`, not the device.